File: c++/leetcode/107. Binary Tree Level Order Traversal II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
  vector<vector<int>> levelOrderBottom(TreeNode *root)
  {
    vector<vector<int>> result;
    if (!root)
      return result;
    queue<TreeNode *> visited;
    visited.push(root);
    int size = 1;

    // pop all from stack add to result
    // push popped if children to stack
    while (1)
    {
      vector<int> temp;
      int added = 0;
      while (size)
      {
        TreeNode *removed = visited.front();
        visited.pop();
        temp.push_back(removed->val);
        if (removed->left)
        {
          visited.push(removed->left);
          added++;
        }
        if (removed->right)
        {
          visited.push(removed->right);
          added++;
        }
        size--;
      }
      result.push_back(temp);
      temp.clear();
      size = added;
      added = 0;
      if (size == 0)
        break;
    }
    reverse(result.begin(), result.end());
    return result;
  }
};

File: javascript/leetcode/107. Binary Tree Level Order Traversal II.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrderBottom = function (root) {
  const result = [];
  if (!root) return result;

  let visited = [];
  visited.push(root);
  let size = 1;

  while (1) {
    let temp = [];
    let added = 0;
    while (size) {
      let removed = visited.shift();
      temp.push(removed.val);
      if (removed.left) {
        visited.push(removed.left);
        added++;
      }
      if (removed.right) {
        visited.push(removed.right);
        added++;
      }
      size--;
    }
    result.push(temp);
    temp = [];
    size = added;
    added = 0;
    if (size === 0) break;
  }
  return result.reverse();
};
